### DISCUSSION TITLE
Crescendo Skipping Stages by MIDI Control Change Events?

### DISCUSSION THREAD

**Discussion by hnb2907** [GrandOrgue/grandorgue#1867] [General]
Hi,

So I've got a Rodgers C445 console, with 2 swell pedals.  I am using 1 of the swell pedals to control the GO Crescendo.  
The Rodgers sends MIDI "Control Change" messages on a channel, with data=0 for swell/crescendo minimum, and value=127 for maximum.

The pedal is a potentiometer, sampled by an ADC inside the organ.

At the moment I am using GO version 3.13.3-1

If I open/close the crescendo pedal slowly, I can see MIDI control change messages, with the data value counting up/down in +1 or -1 steps.  At this time, the crescendo behaviour on the stops is as expected.

However, if I open/close the swell pedal very fast, the sampling speed causes the Rodgers to send only a non-consecutive subset of data values.  The subset of values are in the correct sequence, but with some intermediate values not sent in messages.  When this happens, it appears the GO also skips out some of the intermediate crescendo stop switching.  

At the Rodgers side:  I expect the ADC sampling speed is not sufficient to sample and send a MIDI message for every +/-1 change on the pedal's potentiometer, when the pedal is moving very fast.  And it also probably helps to reduce MIDI choke.

At the GO side:  I "think" it is expecting to see a MIDI message for every +1 or -1 data value in sequence when the pedal is moved?

Here is a simplified example, assuming crescendo has only 10 steps:

1) Starting with "General Cancel" all stops are off.   Moving the crescendo slowly from 0 to 10, all +1 MIDI control change incrementing messages are transmitted from Rodgers to GO, and in GO all crescendo controlled stops are switched on as expected.
2) Next, moving pedal very quickly from 10 to 0, only MIDI messages for data values 7, 5, 3 and 0 are transmitted from Rodgers to GO.  Not all of the stops switch off as expected, some remain on, even at data=0.  It appears that some/all stops associated with missing messages 9, 8, 6, 4, 2 and 1 incorrectly remain on.

Similar behaviour also happens when going fast from 0 to 10, where some stops would not be switched on.

I think that this is what is happening?  Could one of the developers confirm that my diagnosis is correct please?
If it is, I believe that GO should internally trigger in sequence, any missing crescendo steps (e.g. here 9, 8, 6, 4, 2 and 1), where the intermediate MIDI messages were not present.

Cheers,
Chris.

**Comment by hnb2907** [comment]
Moved from Developers to General, probably I created this thread in the wrong place

**Reply by oleg68** [reply]
I suggest you to setup the crescendo in the way that every step would contain all the stops from all previous ones plus some extra stops. In this case even some intermediate steps are missed, you will enable/disable all the stops.
